File: Retaguarda/Torneio.Domain/Entities/ParcelaTorneio.cs
using Torneio.Domain.Enums;

namespace Torneio.Domain.Entities;

public class ParcelaTorneio
{
    public Guid Id { get; private set; }
    public Guid TorneioId { get; private set; }
    public Guid MembroId { get; private set; }
    public TipoParcelaTorneio TipoParcela { get; private set; }
    public int NumeroParcela { get; private set; }
    public Guid? ReferenciaId { get; private set; }
    public string Descricao { get; private set; } = null!;
    public decimal Valor { get; private set; }
    public DateTime Vencimento { get; private set; }
    public bool VencimentoEditadoManual { get; private set; }
    public bool Pago { get; private set; }
    public DateTime? DataPagamento { get; private set; }
    public string? Observacao { get; private set; }
    public string? ComprovanteNomeArquivo { get; private set; }
    public DateTime? ComprovanteDataUpload { get; private set; }
    public string? ComprovanteUsuarioNome { get; private set; }
    public string? ComprovanteUrl { get; private set; }
    public string? ComprovanteContentType { get; private set; }

    private ParcelaTorneio() { }

    public static ParcelaTorneio Criar(
        Guid torneioId,
        Guid membroId,
        TipoParcelaTorneio tipoParcela,
        int numeroParcela,
        string descricao,
        decimal valor,
        DateTime vencimento,
        Guid? referenciaId = null)
    {
        return new ParcelaTorneio
        {
            Id = Guid.NewGuid(),
            TorneioId = torneioId,
            MembroId = membroId,
            TipoParcela = tipoParcela,
            NumeroParcela = numeroParcela,
            ReferenciaId = referenciaId,
            Descricao = descricao.Trim(),
            Valor = valor,
            Vencimento = vencimento,
            VencimentoEditadoManual = false,
            Pago = false
        };
    }

    public void AtualizarValor(decimal valor)
    {
        Valor = valor;
    }

    public void AtualizarDescricao(string descricao)
    {
        Descricao = descricao.Trim();
    }

    public void AtualizarVencimento(DateTime vencimento, bool editadoManual = true)
    {
        Vencimento = vencimento;
        VencimentoEditadoManual = editadoManual;
    }

    public void AtualizarObservacao(string? observacao)
    {
        Observacao = string.IsNullOrWhiteSpace(observacao) ? null : observacao.Trim();
    }

    public void MarcarComoPago(DateTime? dataPagamento = null)
    {
        Pago = true;
        DataPagamento ??= dataPagamento ?? DateTime.UtcNow;
    }

    public void DesmarcarPagamento()
    {
        Pago = false;
        DataPagamento = null;
    }

    public void AtualizarComprovante(
        string nomeArquivo,
        string url,
        string? contentType,
        string usuarioNome,
        DateTime? dataUpload = null)
    {
        ComprovanteNomeArquivo = nomeArquivo;
        ComprovanteUrl = url;
        ComprovanteContentType = contentType;
        ComprovanteUsuarioNome = usuarioNome;
        ComprovanteDataUpload = dataUpload ?? DateTime.UtcNow;
    }
}

File: App/torneio_app/lib/core/models/parcela_torneio.dart
import 'package:intl/intl.dart';

class ParcelaTorneio {
  final String id;
  final String torneioId;
  final String membroId;
  final String nomeMembro;
  final String tipoParcela;
  final String descricao;
  final int numeroParcela;
  final double valor;
  final DateTime vencimento;
  final bool vencimentoEditadoManual;
  final bool pago;
  final DateTime? dataPagamento;
  final String? observacao;
  final bool inadimplente;
  final String? comprovanteNomeArquivo;
  final DateTime? comprovanteDataUpload;
  final String? comprovanteUsuarioNome;
  final String? comprovanteUrl;
  final String? comprovanteContentType;

  const ParcelaTorneio({
    required this.id,
    required this.torneioId,
    required this.membroId,
    required this.nomeMembro,
    required this.tipoParcela,
    required this.descricao,
    required this.numeroParcela,
    required this.valor,
    required this.vencimento,
    required this.vencimentoEditadoManual,
    required this.pago,
    this.dataPagamento,
    this.observacao,
    required this.inadimplente,
    this.comprovanteNomeArquivo,
    this.comprovanteDataUpload,
    this.comprovanteUsuarioNome,
    this.comprovanteUrl,
    this.comprovanteContentType,
  });

  factory ParcelaTorneio.fromJson(Map<String, dynamic> json) => ParcelaTorneio(
        id: json['id'] as String,
        torneioId: json['torneioId'] as String,
        membroId: json['membroId'] as String,
        nomeMembro: json['nomeMembro'] as String? ?? '',
        tipoParcela: json['tipoParcela'] as String? ?? '',
        descricao: json['descricao'] as String? ?? '',
        numeroParcela: json['numeroParcela'] as int? ?? 0,
        valor: (json['valor'] as num?)?.toDouble() ?? 0,
        vencimento: DateTime.parse(json['vencimento'] as String),
        vencimentoEditadoManual: json['vencimentoEditadoManual'] as bool? ?? false,
        pago: json['pago'] as bool? ?? false,
        dataPagamento: json['dataPagamento'] != null
            ? DateTime.tryParse(json['dataPagamento'] as String)
            : null,
        observacao: json['observacao'] as String?,
        inadimplente: json['inadimplente'] as bool? ?? false,
        comprovanteNomeArquivo: json['comprovanteNomeArquivo'] as String?,
        comprovanteDataUpload: json['comprovanteDataUpload'] != null
            ? DateTime.tryParse(json['comprovanteDataUpload'] as String)
            : null,
        comprovanteUsuarioNome: json['comprovanteUsuarioNome'] as String?,
        comprovanteUrl: json['comprovanteUrl'] as String?,
        comprovanteContentType: json['comprovanteContentType'] as String?,
      );

  String get valorFormatado => NumberFormat.currency(locale: 'pt_BR', symbol: 'R\$').format(valor);
}
